### DISCUSSION TITLE
Feature Suggestions

### DISCUSSION THREAD

**Discussion by willmcgugan** [Textualize/textual#3] [Ideas]
If there is a feature you really want to see in Rich TUI, post it here!

**Comment by graingert** [comment]
I'd love to see trio support

**Reply by willmcgugan** [reply]
I'm looking in to AnyIO as an abstraction layer.

**Reply by graingert** [reply]
Thanks, let us know if there's anything you need. Eg, I'm looking at a PriorityQueue version of MemoryObjectStream

**Reply by willmcgugan** [reply]
I didn't realise you were an AnyIO dev! I probably would need a priority queue or equivalent.

Other than that my async needs should be quite modest. It doesn't even need to do any IO, although applications build with rich.tui probably will.

**Reply by lovelydinosaur** [reply]
I'd be motivated to help `textual` bring in trio support.
Is this something you've crossed paths with much @willmcgugan / any existing motivation here without external prodding?

(My ideal stack here is being able to run neato raspberry pi devices with nifty `textual` apps, with `trio` for concurrency, and `httpx` for networkingy bits.)

**Reply by willmcgugan** [reply]
Things have moved on a lot since this post!

We're in a better position to tackle this, now that the internals have stabilized. In all honesty though it feels low priority. There are a lot of things that folk are asking for in Textual, but very few are asking for this. And my concern is that we break things for users that don't care about such things.

If it could be done in a way that wouldn't impact the user experience and performance in any meaningful way, I wouldn't have any objections. Especially if you were helping, Tom!

**Reply by lovelydinosaur** [reply]
> And my concern is that we break things for users that don't care about such things.

Your exposure here looks fairly modest, a non-API impacting switch to `anyio` is probably realistic.

> In all honesty though it feels low priority. There are a lot of things that folk are asking for in Textual, but very few are asking for this.

This doesn't surprise me.
Nonetheless I'd advocate strongly for `trio` to anyone looking at async concurrency in Python.

(The trade-off in ecosystem support is the one sticking point.)

**Reply by willmcgugan** [reply]
No objection then.

There was an [attempt](https://github.com/Textualize/textual/pull/1291), but I think Textual was moving too fast to keep up at that point.

BTW I was assuming you are advocating using AnyIO to get trio support?

**Reply by lovelydinosaur** [reply]
> BTW I was assuming you are advocating using AnyIO to get trio support?

In this instance, yes.

**Reply by Zac-HD** [reply]
hey @willmcgugan - I'm also keen to use Textual on Trio these days.  Just wanted to check that you'd still be happy to get some PRs porting things over to `anyio`?  I'm confident we can avoid breaking any current users 🙂

**Reply by willmcgugan** [reply]
Now would be a good time. Textual is stable, and the internals haven't changed in quite a while.

If you are confident backwards compatibility can be maintained, and there will be no impact in performance, I'm all for it!

**Reply by davidbrochart** [reply]
Is there anybody on it already? We should probably coordinate in order to avoid duplicating efforts.

**Comment by wireboy5** [comment]
Have you considered an API similar to FastAPI or flask? I understand that those are web servers, but an API like that may fit this project. For example, a refractor of examples/simple.py:

```python
from rich.markdown import Markdown

from textual import events
from textual.app import Textual
from textual.widgets.header import Header
from textual.widgets.placeholder import Placeholder
from textual.widgets.window import Window

with open("richreadme.md", "rt") as fh:
    readme = Markdown(fh.read(), hyperlinks=True, code_theme="fruity")

app = Textual()

app.bind({
    "q": "quit",
    "ctrl+c": "quit",
    "b":"view.toggle('left')"
})

@app.on("startup")
async def on_startup(event: events.Startup) -> None:
    await app.view.mount_all(
        header=Header(app.title), left=Placeholder(), body=Window(readme)
    )



app.run()
```

This could also be implemented to work like so:
```python
from rich.markdown import Markdown

from textual import events
from textual.app import Textual
from textual.widgets.header import Header
from textual.widgets.placeholder import Placeholder
from textual.widgets.window import Window

with open("richreadme.md", "rt") as fh:
    readme = Markdown(fh.read(), hyperlinks=True, code_theme="fruity")

class App(Textual):
    def __init__(self):
        self.bind({
            "q": "quit",
            "ctrl+c": "quit",
            "b":"view.toggle('left')"
        })
    
    @on("startup")
    async def on_startup(self, event: events.Startup) -> None:
        await self.view.mount_all(
            header=Header(self.title), left=Placeholder(), body=Window(readme)
        )

app = App()

app.run()
```

This would just help to make things less hard-coded and more dynamic. For example: if the user wanted to use `app.on_startup` for something else.

I was also thinking maybe use a bind function instead of the KEYS variable that can also bind at runtime, in three variants:
```python
# Dictionary of key to event
app.bind({
    "q": "quit",
    "ctrl+c": "quit",
    "b":"view.toggle('left')"
})

# List to events
app.bind(["q","ctrl+c"], "quit")

# Key to event
app.bind("b", "view.toggle('left')")
```

This would make adding and removing keybinds at runtime much easier. (At least, that is what I feel)

All in all, I am very excited to see this project, and wish you much luck and happy coding. 😃

**Reply by wireboy5** [reply]
Oh! I also almost forgot. I do not know if it already works like this, but is there/will there be support for mounting textual apps inside of other textual apps?

Also, I know most of this is polishing stuff, but these were the first things that came to mind when I looked at the project. 👍

**Reply by willmcgugan** [reply]
I prefer the Class based API for this. Events don't quite map to routes in a web server. Routes can grow indefinitely, but there will only ever be a finite number of events.

I might implement a bind function, although wouldn't expect it to be used much. Binding is rarely that dynamic. You set it once on startup and never change it. The exception would be if you are loading bindings from a configuration file on startup.

I considered a decorator approach, but decided against it. I don't see much benefit in having an event handler named in a different way. What would you call a key handler other than "on_key"? Ditto "on_startup". Adding `@on("startup")` just creates repetition. I think the "convention over configuration" approach works better here.

I hadn't considered mounting apps inside apps. It should be possible though. I can already think of some interesting use cases, so thanks for the suggestion!

**Reply by wireboy5** [reply]
Thanks for the reply!

> Routes can grow indefinitely, but there will only ever be a finite number of events.

What do you think about user defined events? And keys can be bound to these events? So instead of using `"view.toggle('left')"` you could use `"toggle_left_panel"`. Just a thought.

> I considered a decorator approach, but decided against it. I don't see much benefit in having an event handler named in a different way. What would you call a key handler other than "on_key"? Ditto "on_startup". Adding `@on("startup")` just creates repetition. I think the "convention over configuration" approach works better here.

I agree, it just sprang to mind because I have seen some other frameworks use this format.

Awesome project, and good luck!

**Reply by willmcgugan** [reply]
The action system allows for custom actions, which lets you bind keys to methods beginning with `action_`. For instance, here's the view.toggle action:

```python
    async def action_toggle(self, layout_name: str) -> None:
        visible = self.layout[layout_name].visible
        self.layout[layout_name].visible = not visible
        await self.layout_update()
```

but you could also define an action on app like this:

```
async def action_toggle_left_panel(self) -> None:
    ...
```

I'm going to flesh out that system, and allow widgets to emit events that can be caught by a parent widget.

**Reply by delulu3i4** [reply]
To add to discussion in terms of binding, one way that could think of dynamic binding is if app allowed to changed the shortcuts while using app (rather just configuring for app start up). Obviously Likely fine if change took affect when restarting app, but being able to do in session could be nice

**Comment by 1dancook** [comment]
Would love to see a scrollable list view. But, to make it really awesome the implementation hopefully would be something like Rich tables.

Along with that, a few thoughts:

- being able to define what the visual selection looks like
- sticky headers for columns?
- methods and properties to interact with the selection
- I would expect that apart from scrolling the list with keys, I can use some defined key mappings to do some function with the selection

**Reply by willmcgugan** [reply]
Absolutely, tables are a killer feature.

**Reply by Cvaniak** [reply]
I was looking for scrollable list view and as I see this is in `Backlog` now. So for anyone who will be looking for something in the mid time I created `hacky` widget [here](https://github.com/Cvaniak/TextualListViewUnofficial).
It is not in pip because I think that official solution will be much better. I also have not create fork because I don't think that I understand `Textual` enough to write clean and style consistent code with rest of the project. (but if it will help I can try)

**Comment by P403n1x87** [comment]
I'd love to see support for resourceful TUIs in ways similar to https://p403n1x87.github.io/the-austin-tui-way-to-resourceful-text-based-user-interfaces.html or https://github.com/salt-die/nurses so that I don't have to maintain my own framework.

**Comment by Etzeitet** [comment]
I would love to see ready to use widgets, such as Menus, Forms (possibly with validation hooks), and popup dialogs. 

We've been using npyscreen as it has those features, but that package hasn't been updated in years and we're looking to move away from it. rich-tui is looking very promising!

**Reply by willmcgugan** [reply]
Absolutely. There will eventually be a builtin library of widgets to cover most use cases.

**Comment by timfi** [comment]
Hi Will, I posted this as an Issue before I noticed this discussion… 😅

One widget I commonly use in many different contexts is some sort of _tabbed container_. I'm willing to open a PR for this myself, but I thought it might be nice to first discuss the API for it.

I'd expect the following usage pattern:
```python
class MyApp(App):
    async def on_mount(self, event: events.Mount) -> None:
        tab1 = Tab("First Tab Label")
        # ... add some widgets to tab1
        await tab1.view.dock(someWidget())
        await tab1.view.dock(someWidget())
        await tab1.view.dock(someWidget())
    
        tab2 = Tab("Second Tab Label")
        # ... add some widgets to the tab
        await tab2.view.dock(someWidget())
        await tab2.view.dock(someWidget())
        await tab2.view.dock(someWidget())
        
        tabs = Tabs([tab1, tab2])
        await self.view.dock(tabs)
```

And I'd expect the output to look similar to this:
```
┌─────────────────┬──────────────────┬─────┄┄┄┄┄┄┄┄┄┄┄────┐
│ First Tab Label │ Second Tab Label │                    │
├─────────────────┴──────────────────┴─────┄┄┄┄┄┄┄┄┄┄┄────┤
│                                                         │
│         [ Content of selected Tab ]                     │
┆                                                         ┆
┆                                                         ┆
│                                                         │
└──────────────────────────────────────────┄┄┄┄┄┄┄┄┄┄┄────┘
```
_With the small difference of the selected being highlighted somehow, be it bold or underlined or italic or whatever._

What are your thoughts on something like this?

**Reply by willmcgugan** [reply]
Hi Tim,

Absolutely, Textual needs something along those lines. You're welcome to tackle it, but I'm still to implement a general solution for switching content in / out. You can do it now by overlaying two widgets over each other and setting their visible flags. But the interface is a little clunky. You might want to wait for that.

In calculator.py there is an example of this switching. The AC button will appear change to C, but it is two distinct buttons in code.

**Reply by timfi** [reply]
Good to know that this will be made easier in the future! But I think I'll have a go at it now (even if just for the challenge… 😅 ) and if I get something satisfactory I'll drop a PR for it. 😄

**Reply by timfi** [reply]
Alrighty! I got something of the ground that looks half decent. 🥳 I'll try and clean it up a bit and open a PR for some discussion on it some time tomorrow. 😴

**Reply by timfi** [reply]
@willmcgugan Not quite done refining things yet, but for a first look you can have a look at this commit: https://github.com/timfi/textual/commit/234567ba20571e5a3f989a7490c96352925d0bde

**Reply by timfi** [reply]
![Screenshot 2021-08-11 at 18 50 31](https://user-images.githubusercontent.com/30415153/129070248-467d1de9-23c1-43c5-8b2a-6641f15ff2d1.png)

I'm almost done, the only problem that I am having write now is that whack spacing of the tab handles…
https://github.com/timfi/textual/commit/2f94ea04b7fb3c2a2a16bd48ec036f4387fed27d

**Reply by timfi** [reply]
Done! 🥳

**Comment by USLTD** [comment]
@willmcgugan,
I have a feature suggestion for Textual. I don't know if it exists yet or not but if not why don't add hyperlinks? One which will open hyperlinks on click, like HTML's <a href="LINK"></a>? Or add JavaScript's and HTML's feature onclick for each widget/text/class/etc? Also second suggestion is to add more widgets like Graphs (using ASCII or Braille characters) and Canvas (using Drawille library).
Some TUI libraries written in another languages have them so I think it will be great to add those to Textual too.

**Reply by BillScherer** [reply]
I've used the Sparklines package with Rich, and that works well for simple, yet useful charts; I expect it will work with Textual too, I just haven't tried it yet.

**Reply by willmcgugan** [reply]
Links do exist currently in Textual. You  can mark up text, detect clicks, and handle as required. I'll be expanding on that in the future.

There will be plots and graphs in the future.

@BillScherer yep, if it works in Rich, it will work with Textual.

**Comment by dissolved** [comment]
I'd like to see argument parsing including robust config file support. It will probably be one of the first things I do on my own assuming I move forward with using Textual for my project. I will try to do that in a way such that I can extract into a PR for Textual, but wouldn't want to duplicate effort if there are already plans to do this and/or explicitly wishes to keep that as something to be done outside of this framework.

**Comment by delulu3i4** [comment]
Don't know if this as much feature, but as community grows around textual imagine many will create their own reusable widgets. With this wondering if there any community standards and such we want to set (even if simple things like repo naming conventions, etc)

Thinking like what see from some community aspects like see in https://github.com/streamlit/streamlit where they have like streamlit gallery, component templates, streamlit repo readme badge. Another community additions to derive inspiration from is possibly pytest plugins

**Reply by willmcgugan** [reply]
Yeah, building community will be important. Streamlit have done that very well, so we can certainly follow their lead.

**Reply by delulu3i4** [reply]
@willmcgugan awesome. I wonder if also given the idea of using widget as classes is very close to idea way flutter (https://github.com/flutter/flutter) organizes its apps/code, if there is something to learn from there and that community?

**Comment by USLTD** [comment]
@willmcgugan,
Since [<kbd>textual</kbd>](https://github.com/Textualize/textual) is based on web technologies will you ...
1. ... add any scripting language runtime like <kbd>JavaScript (ECMAScript)</kbd> in browsers?
2. ... add <kbd>HTML</kbd> or any <kbd>XML</kbd> style markup language for easier **TUI** design?

**Reply by willmcgugan** [reply]
No to both. I’ll borrow  web technologies where it makes sense, but Textual is a Python framework.

**Reply by USLTD** [reply]
> No to both. I’ll borrow web technologies where it makes sense, but Textual is a Python framework.

Just thought that you too would implement **2nd** option like some other TUI libraries, for example [**PyTermGUI**](https://github.com/bczsalba/pytermgui)

**Reply by delulu3i4** [reply]
I would assume, though not perfect converting html to markdown and using Rich's markdown rendering may help (eg package https://github.com/matthewwithanm/python-markdownify) though perhaps there is better way to go about this

**Comment by erdurano** [comment]
An image or thumbnail widget:
I previously mentioned to @willmcgugan about my intentions to implement a image widget in textual on twitter. I have some reservations and questions about specifics and thought that mentioning them here would be fruitful.
- Would it be satisfy the needs to implement a solution as demonstrated in [this repo](https://github.com/stefanhaustein/TerminalImageViewer). It uses Unicode block characters and is sufficiently supported in all terminals that I can get my hands on yet (Windows Terminal and Konsole). If we need a pixel perfect solution, what would be the approach to picture consuming and transforming to Renderable type implementation look like?
 - Widget protocol is sufficiently plain and I'm able to use it to implement the widget as is. But do we need a Image renderable type in Rich also? Must the work begin there?
 - Wherever the solution is implemented, image file consumption would require pillow as a dependency. Does addition of such a library create a foreseeable problem? If so, what would be the alternate library and/or solution.
 - ~~Should this implementation be based on the main branch or css?~~ This is obviously irrelevant now 😁

**Comment by samontoforce** [comment]
An terminal prompt being automatically generated based on a JSON schema would be great!

**Comment by mgaitan** [comment]
a `SubprocessWidget`  to be able to replace tmux-like interfaces with ad-hoc textual apps.

**Comment by Salvodif** [comment]
I would like to have two things

1. an input boxcon autocomplete similar to https://github.com/darrenburns/textual-autocomplete
in which I can enter multiple values. The typical use would be to enter tags (for example).

2. The other thing I would like to have is the possibility to link css also in widgets, i.e. a CSS_PATH so that I can keep the files conceptually separate
